src/System.Private.DataContractSerialization/src/System/Runtime/Serialization/ReflectionReader.cs

```diff
@@ -0,0 +1,628 @@
+        private readonly static MethodInfo s_objectToKeyValuePairGetKey = typeof(ReflectionReader).GetMethod(nameof(ObjectToKeyValuePairGetKey), BindingFlags.NonPublic | BindingFlags.Public | BindingFlags.Static);
+        private readonly static MethodInfo s_objectToKeyValuePairGetValue = typeof(ReflectionReader).GetMethod(nameof(ObjectToKeyValuePairGetValue), BindingFlags.NonPublic | BindingFlags.Public | BindingFlags.Static);
+
+        private readonly static Type[] s_arrayConstructorParameters = new Type[] { Globals.TypeOfInt };
+        private readonly static object[] s_arrayConstructorArguments = new object[] { 32 };
+
+        public object ReflectionReadClass(XmlReaderDelegator xmlReader, XmlObjectSerializerReadContext context, XmlDictionaryString[] memberNames, XmlDictionaryString[] memberNamespaces, ClassDataContract classContract)
+        {
+            object obj = CreateObject(classContract);
+            context.AddNewObject(obj);
+            InvokeOnDeserializing(context, classContract, obj);
+
+            ReflectionReadMembers(xmlReader, context, memberNames, memberNamespaces, classContract, ref obj);
+            obj = ResolveAdapterObject(obj, classContract);
+
+            InvokeOnDeserialized(context, classContract, obj);
+
+            return obj;
+        }
+
+        public object ReflectionReadCollection(XmlReaderDelegator xmlReader, XmlObjectSerializerReadContext context, XmlDictionaryString collectionItemName, XmlDictionaryString collectionItemNamespace, CollectionDataContract collectionContract)
+        {
+            return ReflectionReadCollectionCore(xmlReader, context, collectionItemName, collectionItemNamespace, collectionContract);
+        }
+
+        public void ReflectionReadGetOnlyCollection(XmlReaderDelegator xmlReader, XmlObjectSerializerReadContext context, XmlDictionaryString collectionItemName, XmlDictionaryString collectionItemNamespace, CollectionDataContract collectionContract)
+        {
+            object resultCollection = context.GetCollectionMember();
+            if (ReflectionReadSpecialCollection(xmlReader, context, collectionContract, resultCollection))
+            {
+                return;
+            }
+
+            if (xmlReader.IsStartElement(collectionItemName, collectionItemNamespace))
+            {
+
+                if (resultCollection == null)
+                {
+                    XmlObjectSerializerReadContext.ThrowNullValueReturnedForGetOnlyCollectionException(collectionContract.UnderlyingType);
+                }
+
+                string itemNameStr = collectionContract.ItemName;
+                string itemNsStr = collectionContract.StableName.Namespace;
+
+                ReflectionReadValueDelegate reflectionReadValueDelegate = null;
+                CollectionSetItemDelegate collectionSetItemDelegate = null;
+                for (int index = 0; index < int.MaxValue; index++)
```

Comment:
>for (int index = 0; index < int.MaxValue; index++) [](start = 16, length = 50)

You can't put more than (int.MaxValue - 1048576) into a collection as that's the maximum index value allowed in an array. See the [Remarks section on System.Array](https://msdn.microsoft.com/en-us/library/system.array(v=vs.110).aspx).  This means we'd fail for some other reason before we exited the loop due to the loop check condition evaluating as true. Turn this into `while(true)` with the index declared and initialized outside of the loop. Post-increment the index value on line 82 and remove the decrement on line 95. This will remove an unnecessary condition evaluation on each iteration.